src/main/java/org/team1540/robot2025/subsystems/drive/Drivetrain.java

```diff
@@ -0,0 +1,424 @@
+
+    private static final LoggedTunableNumber headingKP = new LoggedTunableNumber("Drivetrain/Heading/kP", 6.5);
+    private static final LoggedTunableNumber headingKI = new LoggedTunableNumber("Drivetrain/Heading/kI", 0.0);
+    private static final LoggedTunableNumber headingKD = new LoggedTunableNumber("Drivetrain/Heading/kD", 0.0);
+
+    private final GyroIO gyroIO;
+    private final GyroIOInputsAutoLogged gyroInputs = new GyroIOInputsAutoLogged();
+    private final Module[] modules = new Module[4]; // FL, FR, BL, BR
+    private final SwerveDriveKinematics kinematics = RobotState.getInstance().getKinematics();
+
+    private Rotation2d fieldOrientationOffset = Rotation2d.kZero;
+
+    private Rotation2d rawGyroRotation = Rotation2d.kZero;
+    private SwerveModulePosition[] lastModulePositions = new SwerveModulePosition[4]; // For odometry delta filtering
+    private double lastOdometryUpdateTime = 0.0;
+
+    @AutoLogOutput(key = "Drivetrain/DesiredSpeeds")
+    private ChassisSpeeds desiredSpeeds = new ChassisSpeeds();
+
+    private SwerveSetpoint lastSetpoint;
+    private final SwerveSetpointGenerator setpointGenerator =
+            new SwerveSetpointGenerator(ROBOT_CONFIG, MAX_STEER_SPEED_RAD_PER_SEC);
+
+    private boolean isFFCharacterizing = false;
+    private double ffCharacterizationInput = 0.0;
+
+    private final TrajectoryController trajectoryController = new TrajectoryController(
+            translationKP.get(),
+            translationKI.get(),
+            translationKD.get(),
+            headingKP.get(),
+            headingKI.get(),
+            headingKD.get());
+
+    private final ProfiledPIDController headingController = new ProfiledPIDController(
+            headingKP.get(),
+            headingKI.get(),
+            headingKD.get(),
+            new TrapezoidProfile.Constraints(MAX_ANGULAR_SPEED_RAD_PER_SEC, MAX_ANGULAR_ACCEL_RAD_PER_SEC2));
+
+    private final Alert gyroDisconnected = new Alert("Gyro is disconnected", Alert.AlertType.kError);
+
+    public Drivetrain(
+            GyroIO gyroIO, ModuleIO flModuleIO, ModuleIO frModuleIO, ModuleIO blModuleIO, ModuleIO brModuleIO) {
+        if (hasInstance) throw new IllegalStateException("Instance of drivetrain already exists");
+        hasInstance = true;
+
+        this.gyroIO = gyroIO;
+        modules[0] = new Module(flModuleIO, Module.MountPosition.FL, TunerConstants.FrontLeft);
+        modules[1] = new Module(frModuleIO, Module.MountPosition.FR, TunerConstants.FrontRight);
+        modules[2] = new Module(blModuleIO, Module.MountPosition.BL, TunerConstants.BackLeft);
+        modules[3] = new Module(brModuleIO, Module.MountPosition.BR, TunerConstants.BackRight);
+
+        for (int i = 0; i < 4; i++) {
+            lastModulePositions[i] = modules[i].getPosition();
+        }
+
+        headingController.setTolerance(Math.toRadians(1.0));
+        headingController.enableContinuousInput(-Math.PI, Math.PI);
+
+        AutoBuilder.configure(
```

Comment:
why here and not autos class?

Reply:
Because AutoBuilder is used for more than autos, like teleop autodrive.